compiler/rustc_incremental/src/assert_module_sources.rs

```diff
@@ -118,9 +118,8 @@ impl<'tcx> AssertModuleSource<'tcx> {
         debug!("mapping '{}' to cgu name '{}'", self.field(attr, sym::module), cgu_name);
 
         if !self.available_cgus.contains(&cgu_name) {
-            let mut cgu_names: Vec<&str> =
-                self.available_cgus.iter().map(|cgu| cgu.as_str()).collect();
-            cgu_names.sort();
+            let cgu_names: Vec<String> =
+                self.available_cgus.items().map(|cgu| cgu.as_str().to_owned()).into_sorted(&());
```

Comment:
i feel like this impl should be extendedd to all types whicih implement `HashStable` https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/stable_hasher/trait.ToStableHashKey.html#impl-ToStableHashKey%3CHCX%3E-for-String

with that we should be able to keep using a `Vec<&str>` here? 🤔

Reply:
also, please use `to_sorted_stable_ord` for every type which implements `StableOrd` itself. `&str` should implement `StableOrd`

Reply:
Whoops, forgot this PR existed! My bad

I tried implementing ToStableHashKey for all HashStable implementers using a `impl<T, HCX> ToStableHashKey<HCX> for T where T: HashStable<HCX>`, but that runs into issues where it conflicts with `HashStable`s that already have an explicit ToStableHashKey implementation. Should I still look into doing this (probably could make a macro to implement this for a bunch of types), or just leave it as is?

As for the `to_sorted_stable_ord`, done 👍. There's a few types that appeared to be missing StableOrd impls, most notably `&str` and the tuples (provided that the elements within the tuple are StableOrd themselves). I went ahead and just added the impl for these types - is that correct?

Reply:
`ToStableHashKey` (especially when implementing it via `HashStable`) is a pretty heavy hammer and I've personally never been very happy about the existence of the `ToStableHashKey` to begin with. So my recommendation would be to not invest into it too much.

> There's a few types that appeared to be missing StableOrd impls, most notably &str and the tuples (provided that the elements within the tuple are StableOrd themselves). I went ahead and just added the impl for these types - is that correct?

Yes, for the two instances you mention that is correct. I'll take a look at other impls shortly.

> Also I'd just like to confirm - is it correct to use unstable sort when calling into_sorted_stable_ord (to_sorted_stable_ord but for UnordSet iterators) with items that are &str? If two &strs are Eq, it should be completely unnoticeable if unstable sort swaps them, correct? Just avoids an allocation ig.

Unstable sort is always correct in two cases:
- When we know that no two elements are equal, e.g. when the sequence to be sorted is the set of keys of a map or set.
- When the items have no observable difference, so that re-ordering them cannot make a difference anywhere. This is the case for `&str` (because we assume that we don't look at verbatim pointer values). Note though, that the entire item has to be considered here: if sorting a sequence of `(&str, SomeOtherType)` with the `&str` as the sort key, then unstable sorting would be allowed to introduce indeterminism.

Reply:
All new impls of `StableOrd` and `ToStableHashKey` look good to me 👍